edgelet/iotedged/src/lib.rs

```diff
@@ -264,49 +280,80 @@ impl Main {
                     let code = start_api(
                         &settings,
                         hyper_client.clone(),
-                        &runtime,
-                        &key_store,
+                        &$runtime,
+                        &$key_store,
                         cfg.clone(),
-                        root_key.clone(),
+                        $root_key.clone(),
                         make_shutdown_signal(),
                         &crypto,
                         &mut tokio_runtime,
                     )?;
                     code == StartApiReturnStatus::Restart
                 } {}
+            }};
+        }
+
+        info!("Provisioning edge device...");
+        match settings.provisioning() {
+            Provisioning::Manual(manual) => {
+                info!("Starting provisioning edge device via manual mode...");
+                let (key_store, provisioning_result, root_key) =
+                    manual_provision(&manual, &mut tokio_runtime)?;
+                start_edgelet!(key_store, provisioning_result, root_key, runtime);
+            }
+            Provisioning::External(external) => {
+                info!("Starting provisioning edge device via external provisioning mode...");
+                let external_provisioning_client =
+                    ExternalProvisioningClient::new(external.endpoint()).context(
+                        ErrorKind::Initialize(InitializeErrorReason::ExternalProvisioningClient),
+                    )?;
+                let external_provisioning = ExternalProvisioning::new(external_provisioning_client);
+
+                let provision_fut = external_provisioning
+                    .provision(MemoryKeyStore::new())
+                    .map_err(|err| {
+                        Error::from(err.context(ErrorKind::Initialize(
+                            InitializeErrorReason::ExternalProvisioningClient,
+                        )))
+                    });
+
+                let provisioning = tokio_runtime.block_on(provision_fut);
+
+                provisioning.and_then(move |prov_result| {
```

Comment:
Might as well do:

```rust
let prov_result = provisioning?;
let credentials = match prov_result.credentials() {
    Some(credentials) => credentials,
    None => { info!(); return Err(...); },
};
if let AuthType::SymmetricKey(symmetric_key) = credentials.auth_type() {
...
```

... instead of introducing an `and_then` closure that just gets `?`'d anyway.